executor/merge_join.go

```diff
@@ -523,10 +523,12 @@ func (e *MergeJoinExec) joinToChunk(chk *chunk.Chunk) (hasMore bool, err error)
 		for ; e.outerIter4Row.Current() != e.outerIter4Row.End(); e.outerIter4Row.Next() {
 			err = e.resultGenerator.emitToChunk(e.outerIter4Row.Current(), e.innerIter4Row, chk)
 			if err != nil || chk.NumRows() == e.maxChunkSize {
-				e.outerIter4Row.Next()
+				if e.innerIter4Row.Current() == e.innerIter4Row.End() {
```

Comment:
We can move this if branch out of the outer if, after `emitToChunk`, then remove `e.outerIter4Row.Next` in line 523 and remove line 530.